### PR TITLE
Gmres core and reference

### PR DESCRIPTION
**Pull request by yanjen**
This is the first pull request for GMRES.

To solve multiple right-hand side GMRES, we need three-dimension matrices to store the Krylov bases and the Hessenberg matrices. In this version, I aligned the vectors in the two-dimensional fashion and use range to access submatrices to do the task. It can be changed back to three-dimension matrix operations once they are implemented.

I need some suggestions on some issues.
- I implemented the apply function with ranges for dense matrices. Since the range is not a linear operator, I cannot use "apply" to call spmv directly and I don't have an idea of how to fix that right now.
- In order to allocate spaces for Krylov bases and the Hessenberg matrix for the GMRES algorithm, a maximum iteration number must be set. Or is there a smarter way to do this?

### CODE REVIEW

**Comment by gflegar** [comment]
Seems that the build is failing for some reason on the server when OpenMP or Reference modules are enabled. Is it working for you?

About your questions:

> I implemented the apply function with ranges for dense matrices. Since the range is not a linear operator, I cannot use "apply" to call spmv directly and I don't have an idea of how to fix that right now.

You could create a dense matrix view from the range, see the [`solve_system` function in `examples/3pt_stencil`](https://github.com/ginkgo-project/ginkgo/blob/develop/examples/3pt_stencil/3pt_stencil.cpp#L177) for details. It won't work for any kind of underlying accessor, as a range is more general than a dense matrix, but it will work if you properly "unpack" `range<accessor::row_major>`. Maybe it would be a good idea to add a static method to `Dense` that takes a range and creates a matrix view from it.

> In order to allocate spaces for Krylov bases and the Hessenberg matrix for the GMRES algorithm, a maximum iteration number must be set. Or is there a smarter way to do this?

Usually people use the "restarted" GMRES (also called `GMRES(k)`) instead of the plain GMRES method. Basically, the difference is that you only keep a history of `k` vectors - which is an extra parameter to the method, and once you compute those `k` vectors, you start over with the initial guess `x_k`. This way the amount of memory stays independent of the number of iterations you want to do, and the trade-off are some new convergence issues. Pages 17 and 18 of [templates](http://www.netlib.org/templates/templates.pdf) describe this in a bit more detail, and also include pseudocode for restarted GMRES.

**Comment by hartwiganzt** [comment]
@gflegar already pointed out the restart issue. We discussed that previously, and I guess we should set a default (e.g. 100), and the user can then modify this default if needed.

**Comment by gflegar** [comment]
Can the default be set in a "smart" way? Templates suggest no...

**Comment by hartwiganzt** [comment]
Please, let's not add another template parameter. This confuses people...

**Comment by gflegar** [comment]
Templates = "Templates for the Solution of Linear Systems: Building Blocks for Iterative Methods", not C++ templates :smile:

**Comment by hartwiganzt** [comment]
@yanjen you may have to update the branch to avoid merge conflicts. Thanks!

**Comment by hartwiganzt** [comment]
Excellent work!

**Comment by yanjen** [comment]
I am still figuring out how to change it into GMRES(m).
But regardless of that, the code should work correctly.

File: core/solver/gmres.hpp [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+namespace gko {
+namespace solver {
+
+
+constexpr int default_max_iter_num = 100;
+
+
+/**
+ * CG or the conjugate gradient method is an iterative type Krylov subspace
```
Comment (hartwiganzt) [inline core/solver/gmres.hpp]: Update text - this is GMRES

File: core/test/solver/gmres.cpp [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+class Gmres : public ::testing::Test {
+protected:
+    using Mtx = gko::matrix::Dense<>;
+    using Solver = gko::solver::Gmres<>;
+    using Big_solver = gko::solver::Gmres<double>;
+
+    Gmres()
+        : exec(gko::ReferenceExecutor::create()),
+          mtx(gko::initialize<Mtx>(
+              {{2, -1.0, 0.0}, {-1.0, 2, -1.0}, {0.0, -1.0, 2}}, exec)),
```
Comment (hartwiganzt) [inline core/test/solver/gmres.cpp]: this is an SPD matrix - better use a non-spd test system

File: reference/test/solver/gmres_kernels.cpp [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+namespace {
+
+
+class Gmres : public ::testing::Test {
+protected:
+    using Mtx = gko::matrix::Dense<>;
+    Gmres()
+        : exec(gko::ReferenceExecutor::create()),
+          mtx(gko::initialize<Mtx>(
+              {{2, -1.0, 0.0}, {-1.0, 2, -1.0}, {0.0, -1.0, 2}}, exec)),
```
Comment (hartwiganzt) [inline reference/test/solver/gmres_kernels.cpp]: same here - let's use a non-spd system.

File: reference/test/solver/gmres_kernels.cpp [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+                                  .with_time_limit(std::chrono::seconds(6))
+                                  .on_executor(exec),
+                              gko::stop::ResidualNormReduction<>::Factory::
+                                  create()
+                                      .with_reduction_factor(1e-15)
+                                      .on_executor(exec))
+                          .on_executor(exec))
+                  .on_executor(exec)),
+          mtx_big(gko::initialize<Mtx>(
+              {{2295.7, -764.8, 1166.5, 428.9, 291.7, -774.5},
```
Comment (hartwiganzt) [inline reference/test/solver/gmres_kernels.cpp]: interesting... but sure - that works!

File: core/solver/gmres.hpp [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+ * method which is suitable for symmetric positive definite methods.
+ *
+ * Though this method performs very well for symmetric positive definite
+ * matrices, it is in general not suitable for general matrices.
+ *
+ * The implementation in Ginkgo makes use of the merged kernel to make the best
+ * use of data locality. The inner operations in one iteration of CG are merged
```
Comment (hartwiganzt) [inline core/solver/gmres.hpp]: GMRES.

**Comment by yanjen** [comment]
@gflegar @hartwiganzt I am working on the restart and preconditioner right now. It might take a few days. Since you have already reviewed this. Would it be better for me to just finish this PR first and create a new one for the restart and preconditioner? What do you think?

File: reference/test/solver/gmres_kernels.cpp (hunk 4 above)
Comment (yanjen) [reply reference/test/solver/gmres_kernels.cpp]: I generated this matrix randomly. If there is a suggestion, I could also change this matrix.

**Comment by gflegar** [comment]
Yes, the preconditioner and restart is another PR. I'm fine with the content of this PR.

File: reference/solver/gmres_kernels.cpp [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+    for (size_type k = 0; k < x->get_size()[1]; ++k) {
+        for (size_type i = 0; i < x->get_size()[0]; ++i) {
+            for (size_type j = 0; j < final_iter_nums->get_const_data()[k];
+                 ++j) {
+                x->at(i, k) +=
+                    Krylov_bases->at(i, j * x->get_size()[1] + k) * y->at(j, k);
+            }
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (yanjen) [reply reference/solver/gmres_kernels.cpp]: But I thought you wanted to gather as many operations as possible to reduce the memory bandwidth? That is what you do in CG and other solvers, right?

[the remark this answers is not in the document]
Comment (gflegar) [reply reference/solver/gmres_kernels.cpp]: Yes, but _function_ is not the same as _kernel_. We want to gather as many operations as possible into a single kernel. Each kernel however, can be composed of multiple functions. The idea is to reduce the amount of kernel launches, as they introduce additional overheads if launched on a different device (e.g. GPU): host-device communication, cache and shared memory flushing, etc.

Functions are more lightweight, as the only overhead is of launching a function (and that's just a jump and storing the parameters on the stack), and the launch will usually be optimized out by the compiler.

For example, if you look into the block-Jacobi preconditioner generation, you'll see that the whole thing is a single kernel. However, it is composed of quite a few functions that make it up.

Bottom line: function calls don't matter for performance with modern compilers (if they are declared `inline`).

File: core/solver/gmres.cpp [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+        dense_b, b_norm.get(), residual.get(), givens_sin.get(),
+        givens_cos.get(), &final_iter_nums, &stop_status,
+        solver::default_max_iter_num));
+    // b_norm = norm(b)
+    // residual = dense_b
+    // givens_sin = givens_cos = 0
+    // final_iter_nums = {0, ..., 0}
+    system_matrix_->apply(neg_one_op.get(), dense_x, one_op.get(),
+                          residual.get());
+    // residual = residual(b) - Ax
```
Comment (pratikvn) [inline core/solver/gmres.cpp]: I think you can write it as `residual = residual - Ax`. The `(b)` makes it a bit confusing.

**Comment by gflegar** [comment]
@yanjen not sure if you saw my answer on you question regarding the kernels:

> Yes, but function is not the same as kernel. We want to gather as many operations as possible into a single kernel. Each kernel however, can be composed of multiple functions. The idea is to reduce the amount of kernel launches, as they introduce additional overheads if launched on a different device (e.g. GPU): host-device communication, cache and shared memory flushing, etc.
> 
> Functions are more lightweight, as the only overhead is of launching a function (and that's just a jump and storing the parameters on the stack), and the launch will usually be optimized out by the compiler.
> 
> For example, if you look into the block-Jacobi preconditioner generation, you'll see that the whole thing is a single kernel. However, it is composed of quite a few functions that make it up.
> 
> Bottom line: function calls don't matter for performance with modern compilers (if they are declared inline).

**Comment by yanjen** [comment]
@gflegar Yes, I am working on it right now 😄 .

**Comment by yanjen** [comment]
@gflegar You can review it again. I have fixed all the problems.
